### PR TITLE
Rebase dev hzet

### PR DESCRIPTION
**Pull request by julamot**
HZETRN code

### CODE REVIEW

File: HZETRN/btrn.py [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+""" 
+Find the vol-id of the geometry that contains the ref_point
+Requires dagmc python library to be loaded
+"""
+def find_ref_vol(ref_point):
+    # code snippet from dagmc.pyx:find_graveyard_inner_box
+    volumes = dagmc.get_volume_list() 
+    for v in volumes:
+        if dagmc.point_in_volume(v, ref_point):
+            return v
```
Comment (makeclean) [inline HZETRN/btrn.py]: what if we are nowhere?

File: HZETRN/btrn.py [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+try:
+   from pyne import material
+   from pyne.material import Material, MaterialLibrary
+except:
+    raise ImportError("The PyNE dependencies could not be imported.")    
+
+try:
+    from pyne import dagmc
+except:
+    raise ImportError("The dagmc dependency could not be imported.")
```
Comment (makeclean) [inline HZETRN/btrn.py]: DAGMC

File: HZETRN/btrn.py [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+#########################################################
+
+import subprocess
+import argparse
+import string
+import os
+import pprint
+
+try:
+   from pyne import material
```
Comment (makeclean) [inline HZETRN/btrn.py]: Material only once

File: HZETRN/btrn.py [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+   from pyne.material import Material, MaterialLibrary
+except:
+    raise ImportError("The PyNE dependencies could not be imported.")    
+
+try:
+    from pyne import dagmc
```
Comment (makeclean) [inline HZETRN/btrn.py]: amalgamate PyNE imports

File: HZETRN/btrn.py [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+    raise ImportError("The dagmc dependency could not be imported.")    
+
+from pyne import data
+from pyne import nucname
+import numpy as np
+import hzetrn as one_d_tool
+import tag_utils
+import logging
+
+
```
Comment (makeclean) [inline HZETRN/btrn.py]: mixed indentation

File: HZETRN/btrn.py [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+spatial_dir = 'spatial'
+
+"""
+    Read the coordinates of the Reference Point from a file.
+    filename = name of file containing triplet
+        If filename = '' the reference point is set
+        to the origin
+"""
+def load_ray_start(filename):
+    # Set the Reference Point.
```
Comment (makeclean) [inline HZETRN/btrn.py]: Move """ comments to inside function name for documentation autogeneration

File: HZETRN/btrn.py [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+filename name of file with unit direction vectors 
+         Its existence is checked before entering this function.
+         The first member of each pair is the polar angle theta.
+	 The second member is the azimuthal angle phi.
+"""
+def load_ray_tuples(filename):
+
+    ray_tuples = []
+    
+    with open(filename) as f:
```
Comment (makeclean) [inline HZETRN/btrn.py]: Check for correct format, fail if can't open or not enough entries

File: HZETRN/btrn.py [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+# 'spatial_dir' is a subdir for storing input files as they are created
+spatial_dir = 'spatial'
+
+"""
+    Read the coordinates of the Reference Point from a file.
+    filename = name of file containing triplet
+        If filename = '' the reference point is set
+        to the origin
+"""
+def load_ray_start(filename):
```
Comment (makeclean) [inline HZETRN/btrn.py]: rename to load_reference_point

File: HZETRN/btrn.py [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+    # Set the Reference Point.
+    ray_start = []
+    if not filename:
+        return [0.0, 0.0, 0.0]
+    else:
+        with open(filename) as f:
```
Comment (makeclean) [inline HZETRN/btrn.py]: What if file doesnt exist/ unreadable

File: HZETRN/btrn.py [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+	 The second member is the azimuthal angle phi.
+"""
+def load_ray_tuples(filename):
+
+    ray_tuples = []
+    
+    with open(filename) as f:
+        for line in f:
+	    nums = map(float, line.split())
+	    z = np.cos(nums[0])
```
Comment (makeclean) [inline HZETRN/btrn.py]: uvw is typical for a direction

File: HZETRN/btrn.py [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+    return np.array(ray_tuples)
+
+
+""" 
+Find the vol-id of the geometry that contains the ref_point
+Requires dagmc python library to be loaded
+"""
+def find_ref_vol(ref_point):
+    # code snippet from dagmc.pyx:find_graveyard_inner_box
+    volumes = dagmc.get_volume_list()
```
Comment (makeclean) [inline HZETRN/btrn.py]: v = false? as initalisation

File: HZETRN/btrn.py (hunk 1 above)
Comment (makeclean) [inline HZETRN/btrn.py]: add break

File: HZETRN/btrn.py [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+           travels from the start point to the graveyard
+
+This depends on dagmc's ray_iterator, which calls ray_fire
+"""
+def slabs_for_ray(start_vol, ref_point, dir, fname_for_vol):
+    is_graveyard = False
+    foundMat = False
+    surf = 0
+    dist = 0
+    huge = 1000000000
```
Comment (makeclean) [inline HZETRN/btrn.py]: can we do like 1.0e308 or something? 1.0e299 - https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=0CB4QFjAA&url=http%3A%2F%2Fstackoverflow.com%2Fquestions%2F3477283%2Fmaximum-float-in-python&ei=Y81LVc2kGdXXoASfjoHICQ&usg=AFQjCNHspk9ywcYFaF8lv9N4BUN3tHIUJw&sig2=8kt0Gvt9Z0uY_GuJYTm7MA&bvm=bv.92765956,d.aWw

File: HZETRN/btrn.py [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+
+    slab_lengths = []
+    slab_mat_names = []
+    vols_traversed = []
+
```
Comment (makeclean) [inline HZETRN/btrn.py]: name functions as verbs

File: HZETRN/btrn.py [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+    # Could be starting in the implicit_complement
+    if start_vol in fname_for_vol:
+        #  Done before we've started
+        if fname_for_vol[start_vol] == 'graveyard':
+	    # print 'Starting volume is graveyard'
+            return slab_length, slab_mat_name 
+    else:
```
Comment (makeclean) [inline HZETRN/btrn.py]: Remove the else: part -> redundant

File: HZETRN/btrn.py [hunk 15]
```diff
@@ -0,0 +1,377 @@
...
+        print 'We are starting in the implicit complement'
+    is_graveyard = False    
+    last_vol = start_vol
+    
+    for (next_vol, dist, surf, xyz) in dagmc.ray_iterator(last_vol, ref_point, dir, yield_xyz=True):
+        # print 'next_vol', next_vol
+        if not is_graveyard and ((dist < huge) and (surf != 0)):
+	    # We have already checked last_vol for being a graveyard
+	    if last_vol in fname_for_vol:
```
Comment (makeclean) [inline HZETRN/btrn.py]: less ambiguous name for fname -> fluka_name, mat_name etc

File: HZETRN/btrn.py [hunk 16]
```diff
@@ -0,0 +1,377 @@
...
+	        slab_lengths.append(dist)
+		# Get a name associated with volume
+	        slab_mat_names.append(fname_for_vol[last_vol])
+		# Optional documentation
+                vols_traversed.append(last_vol)
+
+	    if next_vol in fname_for_vol:
```
Comment (makeclean) [inline HZETRN/btrn.py]: Save the graveyard vol id earlier

Comment (makeclean) [inline HZETRN/btrn.py]: do if next_vol == graveyard : break

File: HZETRN/btrn.py [hunk 17]
```diff
@@ -0,0 +1,377 @@
...
+        '-p', action='store', dest='ray_start', 
+	help='File containing Cartesion coordinate of starting point of all rays')
+
+    """
+    parser.add_argument(
+        '--no_ray_subset', dest='ray_subset', action='store_false',
+        help='Use all rays (default)')
+    parser.set_defaults(ray_subset=False)
+    """
+
```
Comment (makeclean) [inline HZETRN/btrn.py]: read starting coordinate with an arg_parse argument

File: HZETRN/btrn.py [hunk 18]
```diff
@@ -0,0 +1,377 @@
...
+      
+"""
+Argument parsing
+returns : args: -f for the input geometry file, -r for the input ray tuple file
+ref     : DAGMC/tools/parse_materials/dagmc_get_materials.py
+"""
+def parsing():
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument(
```
Comment (makeclean) [inline HZETRN/btrn.py]: short an long forms, e.g. --environment  -e

File: HZETRN/btrn.py [hunk 19]
```diff
@@ -0,0 +1,377 @@
...
+        '-e', action='store', dest='rad_env', 
+	help='The radiation environment: spe (default) or gcr')
+    parser.set_defaults(rad_env='spe')
+
+    parser.add_argument(
+        '-r', action='store', dest='ray_dir_file', 
+	help='The path to the file with ray direction tuples')
+    parser.set_defaults(ray_dir_file='')
+
```
Comment (makeclean) [inline HZETRN/btrn.py]: may want to change help statement for -r and -n, if -r not provided then fire -n random rays, if both fire n ray subset from file

File: HZETRN/btrn.py [hunk 20]
```diff
@@ -0,0 +1,377 @@
...
+    parser.add_argument(
+        '-n', action='store', dest='rand_dirs', 
+	help='The number of random ray directions to use', type=int)
+
+    parser.add_argument(
```
Comment (makeclean) [inline HZETRN/btrn.py]: change to read list instead of file

File: HZETRN/btrn.py [hunk 21]
```diff
@@ -0,0 +1,377 @@
...
+
+    """
+    parser.add_argument(
+        '--no_ray_subset', dest='ray_subset', action='store_false',
+        help='Use all rays (default)')
+    parser.set_defaults(ray_subset=False)
+    """
+
+    parser.add_argument(
+        '-target', action='store', dest='target',
```
Comment (makeclean) [inline HZETRN/btrn.py]: rename to respone material?

File: HZETRN/btrn.py [hunk 22]
```diff
@@ -0,0 +1,377 @@
...
+        help='The target material in which to calculate the response.  \
+	      NB: the target material cross-section must exist'    )
+    parser.set_defaults(target='WATERLIQ')
+
+    args = parser.parse_args()
+
+    if not args.uwuw_file:
```
Comment (makeclean) [inline HZETRN/btrn.py]: probably more arguments are required to run successfully

File: HZETRN/btrn.py [hunk 23]
```diff
@@ -0,0 +1,377 @@
...
+    - Create a temporary, named directory in the current directory
+           to hold a collection of in put files.
+    - Load the given geometry file and get all its FLUKA names
+    - Get all the ray directions we are going to use
+    - Get the reference point and determine the start volume
+    - For each ray direction, starting from the start vol
+      o get the list of materials and distances it passes through
+      o create the contents of the transport input file
+"""
+def main():
```
Comment (makeclean) [inline HZETRN/btrn.py]: main to be broken up into sub functions

File: HZETRN/btrn.py [hunk 24]
```diff
@@ -0,0 +1,377 @@
...
+    global spatial_dir
+    global config_log
```
Comment (makeclean) [inline HZETRN/btrn.py]: these variables are automatically global.

File: HZETRN/btrn.py [hunk 25]
```diff
@@ -0,0 +1,377 @@
...
+
+    # Setup: parse the the command line parameters
+    args = parsing()
+    uwuw_file = os.path.join(os.path.dirname('__file__'), args.uwuw_file)
+
+    cur_path = os.path.dirname(os.path.abspath(__file__)) + '/'
+    run_path = cur_path + args.run_dir + '/'
+    config_path = run_path + config_log
+
+    logging.basicConfig(filename=config_path, level=logging.INFO, format='%(asctime)s %(message)s')
```
Comment (makeclean) [inline HZETRN/btrn.py]: Should not provide logging functionality

File: HZETRN/btrn.py [hunk 26]
```diff
@@ -0,0 +1,377 @@
...
+    message = 'python btrn.py -f ' + args.uwuw_file + \
+              ' -d ' + args.run_dir + \
+              ' -e ' + args.rad_env + \
+              ' -r ' + ('not given' if not args.ray_dir_file else args.ray_dir_file)
+    logging.info(message)
+
+    # Ensure needed template files exist
+    one_d_tool.check_temp_files(run_path)
```
Comment (makeclean) [inline HZETRN/btrn.py]: rename to check_template_file or tmpl files to avoid "temporary" confusion

File: HZETRN/btrn.py [hunk 27]
```diff
@@ -0,0 +1,377 @@
...
+
+    # Transport results for each direction will be placed in this directory:
+    # Ensure it exists before proceeding.
+    data_path = run_path + 'data/'
```
Comment (makeclean) [inline HZETRN/btrn.py]: check if directory already populated?

File: HZETRN/btrn.py [hunk 28]
```diff
@@ -0,0 +1,377 @@
...
+    if not os.path.isdir(data_path):
+	os.mkdir(data_path)
+
+    rad_env_file = data_path + 'rad_env.txt'
```
Comment (makeclean) [inline HZETRN/btrn.py]: come back discuss later

File: HZETRN/btrn.py [hunk 29]
```diff
@@ -0,0 +1,377 @@
...
+    with open(rad_env_file, 'w') as f:
+        f.write(args.rad_env)
+
+    spatial_path = cur_path + spatial_dir + '/'
+    if not os.path.isdir(spatial_path):
+	os.mkdir(spatial_path)
+
+    # Load the DAG object for this geometry
+    rtn = dagmc.load(uwuw_file)
```
Comment (makeclean) [inline HZETRN/btrn.py]: check return code?

File: HZETRN/btrn.py [hunk 30]
```diff
@@ -0,0 +1,377 @@
...
+
+    vol_fname_dict = tag_utils.get_fnames_for_vol(uwuw_file)
```
Comment (makeclean) [inline HZETRN/btrn.py]: use cell_material_assignment from PyNE DAGMC

File: HZETRN/btrn.py [hunk 31]
```diff
@@ -0,0 +1,377 @@
...
+    ray_tuples = get_directions(args.ray_dir_file, args.rand_dirs)
+
+    # The default starting point is 0,0,0
+    ref_point = load_ray_start(args.ray_start)
+    start_vol = find_ref_vol(ref_point)
+    
+    for dir in ray_tuples:
+	###################################
+	# For backscatter:  
+	# Create two lists 'a' and 'b'.
```
Comment (makeclean) [inline HZETRN/btrn.py]: seperate the work of the method into a new fuction

File: HZETRN/btrn.py [hunk 32]
```diff
@@ -0,0 +1,377 @@
...
+    ref_point = load_ray_start(args.ray_start)
+    start_vol = find_ref_vol(ref_point)
+    
+    for dir in ray_tuples:
+	###################################
+	# For backscatter:  
+	# Create two lists 'a' and 'b'.  
+	#    'a' is the list of slab materials and lengths in the given direction.  
+	#    'b' is the list of same items going OPPOSITE (180 degrees)
+	print start_vol, ref_point, "dir", dir
```
Comment (makeclean) [inline HZETRN/btrn.py]: un-needed debug io

File: HZETRN/btrn.py [hunk 33]
```diff
@@ -0,0 +1,377 @@
...
+	slab_lens_a, slab_mat_names_a = slabs_for_ray(start_vol, ref_point,  dir, vol_fname_dict)
+	slab_lens_b, slab_mat_names_b = slabs_for_ray(start_vol, ref_point, -dir, vol_fname_dict)
+
+	# reverse order and add backward dir
+        # slab_lens = slab_lens_a[::-1] + slab_lens_b
+        slab_lens      = slab_lens_b[::-1]      + slab_lens_a
```
Comment (makeclean) [inline HZETRN/btrn.py]: change to a[:-1] for type safety

File: HZETRN/btrn.py [hunk 34]
```diff
@@ -0,0 +1,377 @@
...
+	slab_mat_names = slab_mat_names_b[::-1] + slab_mat_names_a
+	num_slabs_to_ref = len(slab_lens_b)
```
Comment (makeclean) [inline HZETRN/btrn.py]: more clear variable name, number_of_slabs_to_reference_point ?

File: HZETRN/btrn.py [hunk 35]
```diff
@@ -0,0 +1,377 @@
...
+	num_mats = len(slab_mat_names)
```
Comment (makeclean) [inline HZETRN/btrn.py]: key counting on number of slabs OR number of materials

File: HZETRN/btrn.py [hunk 36]
```diff
@@ -0,0 +1,377 @@
...
+        #############################################
+	# Create the transport geometry file contents
+        # Need to have a non-zero number of mats
+	transport_input = []
+	if 0 != num_mats:
+	    effective_num_mats = num_mats
+	    last_mat_name = slab_mat_names[0]
+	    last_slab_points = [0.0, slab_lens[0]]
+
```
Comment (makeclean) [inline HZETRN/btrn.py]: create a function called create_input_file, maybe should live in hztran specific location. make_file text for example

File: HZETRN/btrn.py [hunk 37]
```diff
@@ -0,0 +1,377 @@
...
+	    for n in range(1,num_mats):
```
Comment (makeclean) [inline HZETRN/btrn.py]: loop over slabs rather than materials

File: HZETRN/btrn.py [hunk 38]
```diff
@@ -0,0 +1,377 @@
...
+	    # The last slab got set up but not written
+	    transport_input.append(last_mat_name)
+	    transport_input.append(str(len(last_slab_points)))
+	    slab_points_list = ["{0:.1f}".format(x) for x in last_slab_points]
+	    transport_input.append(", ".join(slab_points_list))
+	        
+	    transport_input.insert(0,(str(effective_num_mats)))
+
+	# No materials for this or the opposite direction: make bogus spatial file 
+	else:
```
Comment (makeclean) [inline HZETRN/btrn.py]: do we have to run this?

File: HZETRN/btrn.py [hunk 39]
```diff
@@ -0,0 +1,377 @@
...
+	    transport_input.append('0')
+	    transport_input.append('No material traversed, either direction')
+	    transport_input.append('2')
+	    transport_input.append('0.0 0.0')
+
+	# Need a newline at end of file
+	transport_input.append('\n')
+
+	dir_string = "{0:.4f}".format(dir[0]) + '_' + \
```
Comment (makeclean) [inline HZETRN/btrn.py]: directory naming by index in file rather than directions? or some arbitrary consecutive integer

File: HZETRN/btrn.py [hunk 40]
```diff
@@ -0,0 +1,377 @@
...
+	             "{0:.4f}".format(dir[1]) + '_' + \
+		     "{0:.4f}".format(dir[2]) 
+
```
Comment (makeclean) [inline HZETRN/btrn.py]: reconsider naming of filenames etc

File: HZETRN/btrn.py [hunk 41]
```diff
@@ -0,0 +1,377 @@
...
+	# ToDo: change the base name here and on the other side
+	if len(ray_tuples) < 20:
+	    spatial_filename = 'spatial_ba' + dir_string + '.dat'
+	    # sslab = []
+            # for d in slab_lengths:
+	    #     sslab.append(format(d,'.6f'))
+	    # print ('dist, mats ', sslab, slab_mat_names)
+            #############################################
+	else:
+	    spatial_filename = 'spatial.dat'
```
Comment (makeclean) [inline HZETRN/btrn.py]: global

File: HZETRN/btrn.py [hunk 42]
```diff
@@ -0,0 +1,377 @@
...
+
+	# Write out the spatial file even for no materials traversed
+	# A local subdirectory is used.  This is currently 
+	# hardcoded to ./spatial, but a temporary directory could also be used.
+        spatial_filepath = spatial_path + spatial_filename
```
Comment (makeclean) [inline HZETRN/btrn.py]: check for name exists?

File: HZETRN/btrn.py (hunk 3 above)
Comment (julry) [inline HZETRN/btrn.py]: Done

File: HZETRN/btrn.py (hunk 4 above)
Comment (julry) [inline HZETRN/btrn.py]: Done

File: HZETRN/btrn.py (hunk 5 above)
Comment (julry) [inline HZETRN/btrn.py]: ? Not sure what this is referring to

File: HZETRN/btrn.py (hunk 8 above)
Comment (julry) [inline HZETRN/btrn.py]: This function has been removed; the reference point is now a command line parameter list

File: HZETRN/btrn.py (hunk 6 above)
Comment (julry) [inline HZETRN/btrn.py]: Done

File: HZETRN/btrn.py (hunk 9 above)
Comment (julry) [inline HZETRN/btrn.py]: with open(filename) throws an exception and exits if the file is not found:
IOError: [Errno 2] No such file or directory: 'doesnotexist'
It seems to me this is exactly how I would want it handled anyway.    Again, this code is going away, though the same construction is used in reading the list of directions.

File: HZETRN/btrn.py (hunk 7 above)
Comment (julry) [inline HZETRN/btrn.py]: Added check that will exit if line cannot be split into *at least* two floats

File: HZETRN/btrn.py (hunk 10 above)
Comment (julry) [inline HZETRN/btrn.py]: Done

File: HZETRN/btrn.py (hunk 1 above)
Comment (julry) [inline HZETRN/btrn.py]: v is now initialized to -1.  The caller handles this case.

File: HZETRN/btrn.py (hunk 13 above)
Comment (julry) [inline HZETRN/btrn.py]: Done

File: HZETRN/btrn.py (hunk 14 above)
Comment (julry) [inline HZETRN/btrn.py]: Reworked

File: HZETRN/btrn.py (hunk 15 above)
Comment (julry) [inline HZETRN/btrn.py]: Reworked

File: HZETRN/btrn.py (hunk 16 above)
Comment (julry) [inline HZETRN/btrn.py]: Reworked

File: HZETRN/btrn.py (hunk 18 above)
Comment (julry) [inline HZETRN/btrn.py]: Done

File: HZETRN/btrn.py (hunk 19 above)
Comment (julry) [inline HZETRN/btrn.py]: Updated.

File: HZETRN/btrn.py (hunk 20 above)
Comment (julry) [inline HZETRN/btrn.py]: Done

File: HZETRN/btrn.py (hunk 17 above)
Comment (julry) [inline HZETRN/btrn.py]: Done

File: HZETRN/btrn.py (hunk 21 above)
Comment (julry) [inline HZETRN/btrn.py]: Done

File: HZETRN/btrn.py (hunk 22 above)
Comment (julry) [inline HZETRN/btrn.py]: True, but the other required args have defaults.

File: HZETRN/btrn.py (hunk 24 above)
Comment (julry) [inline HZETRN/btrn.py]: Done.  One deleted.

File: HZETRN/btrn.py (hunk 25 above)
Comment (julry) [inline HZETRN/btrn.py]: Removed

File: HZETRN/btrn.py (hunk 33 above)
Comment (julry) [inline HZETRN/btrn.py]: Done

File: HZETRN/btrn.py (hunk 41 above)
Comment (julry) [inline HZETRN/btrn.py]: Actually not.  The directory containing 'spatial.dat' or spatial_ba_....dat is the global.

File: HZETRN/btrn.py (hunk 32 above)
Comment (julry) [inline HZETRN/btrn.py]: Removed.

**Pull request closed without merging**